### ISSUE TITLE
Concurrent requests for the same resource can result in cache misses

### ISSUE THREAD

**Issue by asajeffrey** [servo/servo#24166]
If two `GET` requests are processed in quick succession, then the second one can end up being a cache miss, since the code to check whether the cache has been hit uses a read lock at https://github.com/servo/servo/blob/6ed204d692182e58b7b44b5a02cac4e758d70a14/components/net/http_loader.rs#L1011 whereas the code to insert into the cache uses a write lock at https://github.com/servo/servo/blob/6ed204d692182e58b7b44b5a02cac4e758d70a14/components/net/http_loader.rs#L1120

**Comment by asajeffrey** [comment]
IRC discussion here https://mozilla.logbot.info/servo/20190909#c16601392

**Comment by asajeffrey** [comment]
This came up in the context of https://github.com/servo/servo/pull/24148 which makes concurrent GETs more likely.

**Comment by asajeffrey** [comment]
The matching spec issue: https://github.com/whatwg/fetch/issues/932

**Comment by gterzian** [comment]
That's interesting. 

it might be the point that our simple "lock around a hashmap" approach to the cache shows its limitations. 

We could handle this quite easily if the cache was message-passing based, where the cache ran in a worker thread, and each fetch thread would send a message to ask for a cached entry, and block on the reply. The cache could then simply let the fetch block if it knows there is another one doing the fetch at the same time. 

I guess it could be done with locks(and condvars?) without an additional cache worker thread, but that might be much more complicated...

**Comment by jdm** [comment]
I don't think a message-passing-based approach would solve this issue - the spec describes two separate operations (check the cache & store the response) which happen at different times, so unless we add a marker entry to the cache which means "something else has requested this URL and there is no response yet" we're going to always run into this problem.

**Comment by gterzian** [comment]
> so unless we add a marker entry to the cache which means "something else has requested this URL and there is no response yet" we're going to always run into this problem

I agree. What I mean is that it would be easier to add such a marker, as it would any other "state" in addition to the cached entries, if the cache where to run in it's own thread and handle messages, versus adding more shared-state to the current structure.

**Comment by gterzian** [comment]
Ok so the alternative, as mentioned in the IRC, is to simply have an additional map of `HashMap<CacheKey, Arc<(Condvar, Mutex<boo>>`. 

But there is a big if, in the current structure, since in order to call any method on the cache, you need to lock it, then you can't really start blocking on a condvar while holding this lock, since it would block all other fetches trying to use the cache for completely different keys. 

And passing the condvar/mutex pair around the `http_loader` seems ugly, if possible at all, I'd much rather do this internally to the cache. 

So I propose we revive https://github.com/servo/servo/pull/23553/files although a drastically simpler version that doesn't come with all the fine-grained locking I tried then. 

But the idea of having fethes operation on entries, versus the cache as a whole, means that you can put the condvar/mutex in the entries, and you can actually block on it, since it will only affect other fetches using the same entry.

**Comment by gterzian** [comment]
Ok so I think one idea from https://github.com/servo/servo/pull/23553/files could be useful, which is to have fetches operate on entries, versus the cache as a whole. 

Because a problem with the current structure is that we're operation on a `RwLock<HttpCache>`, which means it's going to be pretty hard to hide any sort of condvar based logic inside of it, since you'd need to acquire the outer lock first, and to tie that one to a condvar would require changing the whole structure of `HttpState`. 

Instead, if fetches operate on entries, we can put a condvar inside of it, tie it to a new mutex around a "pending fetch" flag, and just have every other fetch using the same entry block on it, without affecting ALL fetches.

**Comment by gterzian** [comment]
I will take a look at using entries, and a condvar.
